bin/dlang/install

```diff
@@ -14,7 +14,14 @@ case "$DMD" in
     dmd*   ) PKG= ;;
     1.*    ) PKG="dmd1=$DMD-$DIST" ;;
     2.*.s* ) PKG="dmd-transitional=$DMD-$DIST" ;;
-    2.*    ) PKG="dmd-bin=$DMD libphobos2-dev=$DMD" ;;
+    2.*    ) # For some reason &>/dev/null at the next line would result at
+             # always-success status code:
+             if apt-cache show dmd-compiler=$DMD ; then
+                PKG="dmd-compiler=$DMD dmd-tools=$DMD libphobos2-dev=$DMD"
```

Comment:
@leandro-lucarella-sociomantic ok, I see the problem here - this script is being run on the host system which doesn't have D-APT repos added (and can be arbitrary system in general). Docker container is not yet available.

Reply:
Ping @leandro-lucarella-sociomantic

Reply:
arrggghhh! Dammit!

Reply:
Mmmm, weird. Using `apt` locally, if I only specify `dmd-bin=2.077.1-0.1` it does work, while using `dmd-bin=2.077.1-0.1 libphobos2-dev=2.077.1-0.1` doesn't.

Reply:
Mmm, it seems resolution happens once the package was once installed, even if it's currently removed. That really sucks...

Reply:
This is driving me nuts! Because:
```console
$ apt-cache show dmd-bin=2.078.0-0 | grep Depends:
Depends: dmd-compiler (= 2.078.0-0), dmd-tools (= 2.078.0-0)
$ apt-cache show dmd-compiler=2.078.0-0 | grep Depends:
Depends: libc6, libc6-dev, libphobos2-dev (= 2.078.0-0), gcc, libgcc1, libstdc++6
$ apt-cache show libphobos2-dev=2.078.0-0 | grep Depends:
Depends: dmd-compiler (= 2.078.0-0), libphobos2-78 (= 2.078.0-0), pkg-config
$ sudo apt install dmd-bin=2.078.0-0
...
The following packages have unmet dependencies:
 dmd-bin : Depends: dmd-compiler (= 2.078.0-0) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
$ sudo apt install dmd-compiler=2.078.0-0
...
The following packages have unmet dependencies:
 dmd-compiler : Depends: libphobos2-dev (= 2.078.0-0) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
$ # BUT ADDING libphobos2-dev makes both work!
$ sudo apt install dmd-bin=2.078.0-0 libphobos2-dev=2.078.0-0
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  dmd-compiler dmd-tools libcurl3 libphobos2-78 pkg-config
Suggested packages:
  dmd-doc gcc-multilib
The following NEW packages will be installed:
  dmd-bin dmd-compiler dmd-tools libcurl3 libphobos2-78 libphobos2-dev pkg-config
0 upgraded, 7 newly installed, 0 to remove and 4 not upgraded.
Need to get 8355 kB/8587 kB of archives.
After this operation, 113 MB of additional disk space will be used.
Do you want to continue? [Y/n] 
$ sudo apt install dmd-compiler=2.078.0-0 libphobos2-dev=2.078.0-0
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  libcurl3 libphobos2-78 pkg-config
Suggested packages:
  dmd-doc dmd-tools gcc-multilib
The following NEW packages will be installed:
  dmd-compiler libcurl3 libphobos2-78 libphobos2-dev pkg-config
0 upgraded, 5 newly installed, 0 to remove and 4 not upgraded.
Need to get 7436 kB/7667 kB of archives.
After this operation, 106 MB of additional disk space will be used.
Do you want to continue? [Y/n] 
```

* If `dmd-bin` depends on `dmd-tools (= 2.078.0-0)` and `dmd-compiler (= 2.078.0-0)`, why don't we need to specify those explicitly? (when installing `dmd-bin libphobos-dev`)
* If `libphobos2-dev` depends on `libphobos2-78 (= 2.078.0-0)` why don't we have to specify that one explicitly?

It seems like there is some sort of dependency handling. I don't know if the part that works just works by chance or if we are missing something. I couldn't find any useful information online sadly.

Also, after my local tests using just `install dmd-bin=2.077.1-0.1` that worked, I could never reproduce it in docker, not even removing all DMD related packages.

I guess we'll have to switch back to the simplistic solution of just picking the package based on the version and our previous knowledge of how d-apt works. But it keeps bothering me that some dependency handling is working and that we already had a hack (having to add `libphobos2-dev` explicitly), when other packages are handled automatically...

:rage:

Reply:
> I guess we'll have to switch back to the simplistic solution of just picking the package based on the version and our previous knowledge of how d-apt works.

Yeah, I will revert to initial version. My only guess about why for some packages it works is that they have `depends:` statement with exact version (like `== 2.077.1-0`) and other haven an open one (like `>= 2.077`) and apt refuses to try guessing the latter.

Reply:
Nope, `libphobos2-78 (= 2.078.0-0)` is installed automatically.

Reply:
Got dammit!!!

After having a very good theory of why it worked locally (I noticed I had a more recent version of `apt` because I use the repo `xenial-proposed`) and finding out it was bogus (it didn't solve the problem inside docker), I finally **got it**.

**IS THE PACKAGE PINNING**.

Removing:
```
Package: dmd-bin
Pin: version 2.070.*
Pin-Priority: 990

Package: libphobos2-dev
Pin: version 2.070.*
Pin-Priority: 990
```

From `/etc/apt/preferences.d/cachalot` inside docker makes `apt install dmd-bin=2.078.*` work like a charm...

Reply:
I have the feeling this could bite us in many other forms, so maybe it would be good to just disable the pinning when installing dmd in beaver?

Actually removing only `libphobos2-dev` from the pinning is enough. Since it's tied to `dmd-bin` I guess is fine to just remove that?

Reply:
Aaaand, it's not the pinning. That only worked because:

* for `2.078.*` it resolves to the latest version
* for `2.077.*` dependencies are broken (`Depends: dmd-compiler, dmd-tools`) since there is no version specified, the actual compiler being installed is `2.078.0-0`, or the latest one, instead of the requested one, only the dummy `dmd-bin` package will be at `2.077.*`.

Reply:
```console
$ sudo apt install dmd-bin=2.077.*
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Selected version '2.077.1-0.1' (. d-apt:d-apt [amd64]) for 'dmd-bin'
The following additional packages will be installed:
   dmd-compiler (2.078.0-0)
   dmd-tools (2.078.0-0)
   libphobos2-78 (2.078.0-0)
   libphobos2-dev (2.078.0-0)
Suggested packages:
   dmd-doc (2.078.0-0)
The following NEW packages will be installed:
   dmd-bin (2.077.1-0.1)
   dmd-compiler (2.078.0-0)
   dmd-tools (2.078.0-0)
   libphobos2-78 (2.078.0-0)
   libphobos2-dev (2.078.0-0)
0 upgraded, 5 newly installed, 0 to remove and 3 not upgraded.
Need to get 8,355 kB of archives.
After this operation, 112 MB of additional disk space will be used.
Do you want to continue? [Y/n]
```

This can bite people trying to use `2.077.*`.

All points back again to use the dmd version to decide what package to install (installing `dmd-compiler` starting from `2.077.*`).

:grumpy:

PS: I still think `libphobos2-dev shouldn't be pinned though, but it's much more low prio.